File: Python/ch08/MaxDistance.py
class Node:
    def __init__(self, v):
        self.value = v
        self.left = None
        self.right = None


class Info:
    def __init__(self, d, h):
        self.max_distance = d
        self.height = h


def process(r):
    if r is None:
        return Info(True, 0)

    left_info = process(r.left)
    right_info = process(r.right)

    height = max(left_info.height, right_info.height) + 1
    max_distance = max(left_info.max_distance, right_info.max_distance, left_info.height + right_info.height)

    return Info(max_distance, height)


if __name__ == '__main__':
    root = Node(6)
    root.left = Node(4)
    root.right = Node(8)
    root.left.left = Node(2)
    root.left.right = Node(5)
    root.right.left = Node(7)
    root.right.right = Node(9)
    root.left.left.left = Node(1)
    root.left.left.right = Node(3)

    print(process(root).max_distance)

File: Java/src/ch08/MaxDistance.java
package ch08;

public class MaxDistance {

    public static class Node{
        public int value;
        public Node left;
        public Node right;

        public Node(int v){
            value = v;
        }
    }

    public static class Info{
        public int maxDistance;
        public int height;

        public Info(int d,int h){
            maxDistance = d;
            height = h;
        }
    }

    public static Info process(Node root){
        if (null == root){
            return new Info(0,0);
        }

        Info leftInfo = process(root.left);
        Info rightInfo = process(root.right);

        int height = Math.max(leftInfo.height,rightInfo.height) + 1;
        int maxDistance = Math.max(Math.max(leftInfo.maxDistance,rightInfo.maxDistance),leftInfo.height + rightInfo.height);

        return new Info(maxDistance,height);
    }


    public static void main(String[] args) {
        Node root = new Node(6);
        root.left = new Node(4);
        root.right = new Node(8);
        root.left.left = new Node(2);
        root.left.right = new Node(5);
        root.right.left = new Node(7);
        root.right.right = new Node(9);
        root.left.left.left = new Node(1);
        root.left.left.right = new Node(3);

        System.out.println(process(root).maxDistance);
    }

}
